File: apps/mobile/lib/domain/entities/product.dart
class Product {
  final String id;
  final String name;
  final String slug;
  final String description;
  final double retailPrice;
  final double? wholesalePrice;
  final int? wholesaleMinQuantity;
  final int discountPercentage;
  final int stockQuantity;
  final int lowStockThreshold;
  final String? primaryImage;
  final List<String> images;
  final String categoryId;
  final String categoryName;
  final String categorySlug;
  final String sellerId;
  final String? businessName;
  final bool businessVerified;
  final double averageRating;
  final int reviewCount;
  final int soldCount;
  final String status;
  final String createdAt;

  Product({
    required this.id,
    required this.name,
    required this.slug,
    required this.description,
    required this.retailPrice,
    this.wholesalePrice,
    this.wholesaleMinQuantity,
    this.discountPercentage = 0,
    this.stockQuantity = 0,
    this.lowStockThreshold = 5,
    this.primaryImage,
    this.images = const [],
    required this.categoryId,
    required this.categoryName,
    required this.categorySlug,
    required this.sellerId,
    this.businessName,
    this.businessVerified = false,
    this.averageRating = 0,
    this.reviewCount = 0,
    this.soldCount = 0,
    this.status = 'active',
    required this.createdAt,
  });

  factory Product.fromJson(Map<String, dynamic> json) {
    return Product(
      id: json['id'] as String,
      name: json['name'] as String,
      slug: json['slug'] as String,
      description: json['description'] as String? ?? '',
      retailPrice: double.tryParse(json['retail_price']?.toString() ?? '0') ?? 0,
      wholesalePrice: json['wholesale_price'] != null ? double.tryParse(json['wholesale_price'].toString()) ?? 0 : null,
      wholesaleMinQuantity: json['wholesale_min_quantity'] as int?,
      discountPercentage: json['discount_percentage'] as int? ?? 0,
      stockQuantity: json['stock_quantity'] as int? ?? 0,
      lowStockThreshold: json['low_stock_threshold'] as int? ?? 5,
      primaryImage: json['primary_image'] as String?,
      images: (json['images'] as List<dynamic>?)?.map((e) => e is String ? e : (e['image_url'] as String)).toList() ?? [],
      categoryId: json['category_id'] as String? ?? '',
      categoryName: json['category_name'] as String? ?? '',
      categorySlug: json['category_slug'] as String? ?? '',
      sellerId: json['seller_id'] as String? ?? '',
      businessName: json['business'] is Map ? (json['business'] as Map)['business_name'] as String? : null,
      businessVerified: json['business'] is Map ? (json['business'] as Map)['is_verified'] as bool? ?? false : false,
      averageRating: double.tryParse(json['average_rating']?.toString() ?? '0') ?? 0,
      reviewCount: json['review_count'] as int? ?? 0,
      soldCount: json['sold_count'] as int? ?? 0,
      status: json['status'] as String? ?? 'active',
      createdAt: json['created_at'] as String,
    );
  }

  double get effectivePrice => retailPrice * (1 - discountPercentage / 100);
  bool get lowStock => stockQuantity <= lowStockThreshold && stockQuantity > 0;
  bool get outOfStock => stockQuantity <= 0;
}

File: packages/models/src/product.ts
import type { Timestamp, EntityStatus } from './common';

export interface Category {
  id: string;
  name: string;
  slug: string;
  description?: string;
  image_url?: string;
  parent_id?: string;
  order_index: number;
  is_active: boolean;
  created_at: Timestamp;
  updated_at: Timestamp;
}

export interface Subcategory {
  id: string;
  category_id: string;
  name: string;
  slug: string;
  description?: string;
  order_index: number;
  is_active: boolean;
  created_at: Timestamp;
}

export interface Product {
  id: string;
  seller_id: string;
  business_id: string;
  category_id: string;
  subcategory_id?: string;
  name: string;
  slug: string;
  description: string;
  sku: string;
  barcode?: string;
  brand?: string;
  weight?: number;
  weight_unit?: string;
  unit?: string;
  retail_price: number;
  wholesale_price?: number;
  wholesale_min_quantity?: number;
  discount_percentage?: number;
  discount_end_date?: Timestamp;
  stock_quantity: number;
  low_stock_threshold: number;
  is_featured: boolean;
  is_published: boolean;
  status: EntityStatus;
  location?: string;
  region?: string;
  average_rating: number;
  review_count: number;
  sold_count: number;
  created_at: Timestamp;
  updated_at: Timestamp;
}

export interface ProductImage {
  id: string;
  product_id: string;
  image_url: string;
  alt_text?: string;
  is_primary: boolean;
  order_index: number;
  created_at: Timestamp;
}

export interface ProductVideo {
  id: string;
  product_id: string;
  video_url: string;
  thumbnail_url?: string;
  title?: string;
  created_at: Timestamp;
}

export interface Inventory {
  id: string;
  product_id: string;
  quantity_before: number;
  quantity_change: number;
  quantity_after: number;
  change_type: 'purchase' | 'sale' | 'return' | 'adjustment' | 'restock';
  reference_type?: string;
  reference_id?: string;
  notes?: string;
  created_at: Timestamp;
}
